{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c1ce6f55-ecf9-4622-ad49-6532cb05cd89",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# PROXY\n",
    "import subprocess\n",
    "import os\n",
    "\n",
    "result = subprocess.run('bash -c \"source /etc/network_turbo && env | grep proxy\"', shell=True, capture_output=True, text=True)\n",
    "output = result.stdout\n",
    "for line in output.splitlines():\n",
    "    if '=' in line:\n",
    "        var, value = line.split('=', 1)\n",
    "        os.environ[var] = value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f9edd9c1-4f8b-4a55-a54c-ff851b9f383e",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/root/miniconda3/lib/python3.8/site-packages/datasets/load.py:1461: FutureWarning: The repository for bigbio/biored contains custom code which must be executed to correctly load the dataset. You can inspect the repository content at https://hf.co/datasets/bigbio/biored\n",
      "You can avoid this message in future by passing the argument `trust_remote_code=True`.\n",
      "Passing `trust_remote_code=True` will be mandatory to load this dataset from the next major release of `datasets`.\n",
      "  warnings.warn(\n",
      "Repo card metadata block was not found. Setting CardData to empty.\n"
     ]
    }
   ],
   "source": [
    "from datasets import load_dataset\n",
    "datas = load_dataset(\"bigbio/biored\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c105b026-4498-48e2-965d-35b9bada910f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from text2knowledge.strategy1 import extract_concepts\n",
    "\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "10471132-0e44-4beb-959b-49a3226f7aa1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "model_name = 'mistral:latest'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b7f0ac51-d307-4951-856a-eb0a77e18cb4",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " [\n",
      "    {\n",
      "        \"entity\": \"Congenital long QT syndrome (LQTS)\",\n",
      "        \"confidence\": \"5\",\n",
      "        \"category\": \"Disease\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"in utero onset\",\n",
      "        \"confidence\": \"4\",\n",
      "        \"category\": \"Symptom\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"rhythm disturbances\",\n",
      "        \"confidence\": \"3\",\n",
      "        \"category\": \"BiologicalProcess\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"newborn patient\",\n",
      "        \"confidence\": \"5\",\n",
      "        \"category\": \"Anatomy\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"fetal bradycardia\",\n",
      "        \"confidence\": \"4\",\n",
      "        \"category\": \"Symptom\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"2:1 atrioventricular block\",\n",
      "        \"confidence\": \"5\",\n",
      "        \"category\": \"Symptom\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"ventricular tachycardia\",\n",
      "        \"confidence\": \"4\",\n",
      "        \"category\": \"Symptom\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"Na(v)1.5 cardiac sodium channel\",\n",
      "        \"confidence\": \"5\",\n",
      "        \"category\": \"Protein\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"G-->A substitution\",\n",
      "        \"confidence\": \"3\",\n",
      "        \"category\": \"Compound\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"valine (GTG)\",\n",
      "        \"confidence\": \"4\",\n",
      "        \"category\": \"Molecule\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"methionine (ATG)\",\n",
      "        \"confidence\": \"4\",\n",
      "        \"category\": \"Molecule\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"codon 1763\",\n",
      "        \"confidence\": \"3\",\n",
      "        \"category\": \"Compound\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"transmembrane segment 6 of domain IV\",\n",
      "        \"confidence\": \"5\",\n",
      "        \"category\": \"Anatomy\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"persistent tetrodotoxin-sensitive but lidocaine-resistant current\",\n",
      "        \"confidence\": \"4\",\n",
      "        \"category\": \"MolecularFunction\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"positive shift of the steady-state inactivation curve\",\n",
      "        \"confidence\": \"3\",\n",
      "        \"category\": \"BiologicalProcess\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"steeper activation curve\",\n",
      "        \"confidence\": \"3\",\n",
      "        \"category\": \"BiologicalProcess\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"faster recovery from inactivation\",\n",
      "        \"confidence\": \"3\",\n",
      "        \"category\": \"BiologicalProcess\"\n",
      "    }\n",
      "] [\n",
      "    {\n",
      "        \"entity\": \"mu opioid receptor (OPRM1)\",\n",
      "        \"confidence\": \"5\",\n",
      "        \"category\": \"Protein\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"Opioid drugs\",\n",
      "        \"confidence\": \"3\",\n",
      "        \"category\": \"Compound\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"Pain perception\",\n",
      "        \"confidence\": \"5\",\n",
      "        \"category\": \"BiologicalProcess\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"Addiction\",\n",
      "        \"confidence\": \"5\",\n",
      "        \"category\": \"Disease\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"Single nucleotide polymorphism (SNP) A118G\",\n",
      "        \"confidence\": \"5\",\n",
      "        \"category\": \"GeneticVariant\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"N40D substitution\",\n",
      "        \"confidence\": \"3\",\n",
      "        \"category\": \"Mutation\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"allele frequency of 10-32%\",\n",
      "        \"confidence\": \"2\",\n",
      "        \"category\": \"Statistic\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"OPRM1-A118 mRNA allele\",\n",
      "        \"confidence\": \"4\",\n",
      "        \"category\": \"Protein\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"OPRM1-G118 mRNA allele\",\n",
      "        \"confidence\": \"4\",\n",
      "        \"category\": \"Protein\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"Chinese hamster ovary cells\",\n",
      "        \"confidence\": \"5\",\n",
      "        \"category\": \"CellLine\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"actinomycin D\",\n",
      "        \"confidence\": \"3\",\n",
      "        \"category\": \"Compound\"\n",
      "    }\n",
      "] [\n",
      "    {\n",
      "        \"entity\": \"congestive heart failure (CHF)\",\n",
      "        \"confidence\": \"5\",\n",
      "        \"category\": \"Disease\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"anthracyclines\",\n",
      "        \"confidence\": \"5\",\n",
      "        \"category\": \"Compound\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"Childhood Cancer Survivor Study\",\n",
      "        \"confidence\": \"4\",\n",
      "        \"category\": \"Anatomy\" or \"Organism\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"nicotinamide adenine dinucleotide phosphate:quinone oxidoreductase 1 gene (NQO1)\",\n",
      "        \"confidence\": \"5\",\n",
      "        \"category\": \"Gene\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"carbonyl reductase 3 gene (CBR3)\",\n",
      "        \"confidence\": \"5\",\n",
      "        \"category\": \"Gene\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"NQO1*2 polymorphism\",\n",
      "        \"confidence\": \"4\",\n",
      "        \"category\": \"Compound\" or \"Mutation\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"CBR3 V244M polymorphism\",\n",
      "        \"confidence\": \"4\",\n",
      "        \"category\": \"Compound\" or \"Mutation\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"doxorubicin\",\n",
      "        \"confidence\": \"5\",\n",
      "        \"category\": \"Compound\"\n",
      "    }\n",
      "]\n",
      "Note: I assumed that the \"Childhood Cancer Survivor Study\" is a cohort study and it can be categorized as an \"Organism\" or \"Anatomy\" since it refers to a group of individuals who survived cancer. However, it could also be considered as a research study and thus belong to the category \"Research\" or \"Study\". The choice depends on the context and available information.\n",
      "\n",
      "ERROR ### Here is the buggy response:   [\n",
      "    {\n",
      "        \"entity\": \"congestive heart failure (CHF)\",\n",
      "        \"confidence\": \"5\",\n",
      "        \"category\": \"Disease\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"anthracyclines\",\n",
      "        \"confidence\": \"5\",\n",
      "        \"category\": \"Compound\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"Childhood Cancer Survivor Study\",\n",
      "        \"confidence\": \"4\",\n",
      "        \"category\": \"Anatomy\" or \"Organism\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"nicotinamide adenine dinucleotide phosphate:quinone oxidoreductase 1 gene (NQO1)\",\n",
      "        \"confidence\": \"5\",\n",
      "        \"category\": \"Gene\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"carbonyl reductase 3 gene (CBR3)\",\n",
      "        \"confidence\": \"5\",\n",
      "        \"category\": \"Gene\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"NQO1*2 polymorphism\",\n",
      "        \"confidence\": \"4\",\n",
      "        \"category\": \"Compound\" or \"Mutation\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"CBR3 V244M polymorphism\",\n",
      "        \"confidence\": \"4\",\n",
      "        \"category\": \"Compound\" or \"Mutation\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"doxorubicin\",\n",
      "        \"confidence\": \"5\",\n",
      "        \"category\": \"Compound\"\n",
      "    }\n",
      "]\n",
      "Note: I assumed that the \"Childhood Cancer Survivor Study\" is a cohort study and it can be categorized as an \"Organism\" or \"Anatomy\" since it refers to a group of individuals who survived cancer. However, it could also be considered as a research study and thus belong to the category \"Research\" or \"Study\". The choice depends on the context and available information. \n",
      "\n",
      "\n",
      "If you get `404 Client Error: Not Found`, please check if the model name is correct or you have installed the model. If you get `500 Server Error: Internal Server Error`, please check if the model is running.\n",
      "No entities found.\n",
      " [\n",
      "    {\n",
      "        \"entity\": \"metoprolol\",\n",
      "        \"confidence\": \"5\",\n",
      "        \"category\": \"Compound\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"debrisoquine/sparteine type metabolism\",\n",
      "        \"confidence\": \"4\",\n",
      "        \"category\": \"Pathway\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"extensive (EM) metabolizers\",\n",
      "        \"confidence\": \"3\",\n",
      "        \"category\": \"Symptom\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"poor metabolizers (PM)\",\n",
      "        \"confidence\": \"3\",\n",
      "        \"category\": \"Symptom\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"IC50 values\",\n",
      "        \"confidence\": \"4\",\n",
      "        \"category\": \"Metabolite\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"beta-1 adrenoceptor antagonistic potency\",\n",
      "        \"confidence\": \"3\",\n",
      "        \"category\": \"MolecularFunction\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"terbutaline-induced hypokalemia\",\n",
      "        \"confidence\": \"4\",\n",
      "        \"category\": \"Symptom\"\n",
      "    }\n",
      "] [\n",
      "    {\n",
      "        \"entity\": \"Deoxyguanosine kinase (dGK)\",\n",
      "        \"confidence\": \"5\",\n",
      "        \"category\": \"Enzyme (Protein)\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"DGUOK gene\",\n",
      "        \"confidence\": \"5\",\n",
      "        \"category\": \"Gene\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"mitochondrial DNA depletion\",\n",
      "        \"confidence\": \"4\",\n",
      "        \"category\": \"Metabolite\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"hepatocerebral phenotype\",\n",
      "        \"confidence\": \"4\",\n",
      "        \"category\": \"Symptom\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"combined respiratory chain deficiencies\",\n",
      "        \"confidence\": \"3\",\n",
      "        \"category\": \"Disease\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"subtentorial abnormal myelination\",\n",
      "        \"confidence\": \"4\",\n",
      "        \"category\": \"Anatomy\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"moderate hyperintensity in the bilateral pallidi\",\n",
      "        \"confidence\": \"3\",\n",
      "        \"category\": \"Symptom\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"alpha5 helix of the dGK kinase domain\",\n",
      "        \"confidence\": \"4\",\n",
      "        \"category\": \"Protein\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"purine deoxyribonucleosides\",\n",
      "        \"confidence\": \"3\",\n",
      "        \"category\": \"Compound\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"haplotype reconstruction with 10 markers (microsatellites and SNPs)\",\n",
      "        \"confidence\": \"2\",\n",
      "        \"category\": \"MolecularFunction\"\n",
      "    }\n",
      "] [\n",
      "  {\n",
      "    \"entity\": \"Autism spectrum disorders (ASDs)\",\n",
      "    \"confidence\": \"5\",\n",
      "    \"category\": \"Disease\"\n",
      "  },\n",
      "  {\n",
      "    \"entity\": \"anxiety\",\n",
      "    \"confidence\": \"4\",\n",
      "    \"category\": \"Symptom\"\n",
      "  },\n",
      "  {\n",
      "    \"entity\": \"adenosine A(2A) receptor gene (ADORA2A)\",\n",
      "    \"confidence\": \"5\",\n",
      "    \"category\": \"Gene\"\n",
      "  },\n",
      "  {\n",
      "    \"entity\": \"Chromosome 22q11.23\",\n",
      "    \"confidence\": \"4\",\n",
      "    \"category\": \"Anatomy\"\n",
      "  },\n",
      "  {\n",
      "    \"entity\": \"adenosine A(2A) receptor\",\n",
      "    \"confidence\": \"5\",\n",
      "    \"category\": \"Protein\"\n",
      "  },\n",
      "  {\n",
      "    \"entity\": \"caudate nucleus\",\n",
      "    \"confidence\": \"4\",\n",
      "    \"category\": \"Anatomy\"\n",
      "  },\n",
      "  {\n",
      "    \"entity\": \"22q11.2 deletion syndrome\",\n",
      "    \"confidence\": \"5\",\n",
      "    \"category\": \"Disease\"\n",
      "  },\n",
      "  {\n",
      "    \"entity\": \"rs2236624-CC\",\n",
      "    \"confidence\": \"4\",\n",
      "    \"category\": \"Compound\"\n",
      "  },\n",
      "  {\n",
      "    \"entity\": \"rs3761422\",\n",
      "    \"confidence\": \"4\",\n",
      "    \"category\": \"Compound\"\n",
      "  },\n",
      "  {\n",
      "    \"entity\": \"rs5751876\",\n",
      "    \"confidence\": \"4\",\n",
      "    \"category\": \"Compound\"\n",
      "  },\n",
      "  {\n",
      "    \"entity\": \"rs35320474\",\n",
      "    \"confidence\": \"4\",\n",
      "    \"category\": \"Compound\"\n",
      "  },\n",
      "  {\n",
      "    \"entity\": \"phenotypic variability in ASD symptoms\",\n",
      "    \"confidence\": \"3\",\n",
      "    \"category\": \"BiologicalProcess\"\n",
      "  },\n",
      "  {\n",
      "    \"entity\": \"ASD individuals\",\n",
      "    \"confidence\": \"4\",\n",
      "    \"category\": \"Population\"\n",
      "  }\n",
      "] [\n",
      "  {\n",
      "    \"entity\": \"HIV\",\n",
      "    \"confidence\": \"5\",\n",
      "    \"category\": \"Disease\"\n",
      "  },\n",
      "  {\n",
      "    \"entity\": \"Hepatitis B virus (HBV)\",\n",
      "    \"confidence\": \"5\",\n",
      "    \"category\": \"Disease\"\n",
      "  },\n",
      "  {\n",
      "    \"entity\": \"lamivudine (LAM)\",\n",
      "    \"confidence\": \"5\",\n",
      "    \"category\": \"Compound\"\n",
      "  },\n",
      "  {\n",
      "    \"entity\": \"tenofovir (TDF)\",\n",
      "    \"confidence\": \"5\",\n",
      "    \"category\": \"Compound\"\n",
      "  },\n",
      "  {\n",
      "    \"entity\": \"HBsAg\",\n",
      "    \"confidence\": \"4\",\n",
      "    \"category\": \"Protein\"\n",
      "  },\n",
      "  {\n",
      "    \"entity\": \"real-time polymerase chain reaction (PCR)\",\n",
      "    \"confidence\": \"3\",\n",
      "    \"category\": \"Technique\"\n",
      "  },\n",
      "  {\n",
      "    \"entity\": \"HBV genotypes/subgenotypes\",\n",
      "    \"confidence\": \"5\",\n",
      "    \"category\": \"Anatomy\"\n",
      "  },\n",
      "  {\n",
      "    \"entity\": \"antiviral resistance\",\n",
      "    \"confidence\": \"4\",\n",
      "    \"category\": \"BiologicalProcess\"\n",
      "  },\n",
      "  {\n",
      "    \"entity\": \"basal core promoter (BCP)\",\n",
      "    \"confidence\": \"5\",\n",
      "    \"category\": \"Anatomy\"\n",
      "  },\n",
      "  {\n",
      "    \"entity\": \"precore mutations\",\n",
      "    \"confidence\": \"4\",\n",
      "    \"category\": \"Mutation\"\n",
      "  },\n",
      "  {\n",
      "    \"entity\": \"rtL180M + rtM204V\",\n",
      "    \"confidence\": \"5\",\n",
      "    \"category\": \"Mutation\"\n",
      "  },\n",
      "  {\n",
      "    \"entity\": \"rtV173L mutation\",\n",
      "    \"confidence\": \"5\",\n",
      "    \"category\": \"Mutation\"\n",
      "  },\n",
      "  {\n",
      "    \"entity\": \"triple polymerase substitution pattern (rtV173L + rtL180M + rtM204V)\",\n",
      "    \"confidence\": \"5\",\n",
      "    \"category\": \"Mutation\"\n",
      "  },\n",
      "  {\n",
      "    \"entity\": \"sE164D + sI195M\",\n",
      "    \"confidence\": \"5\",\n",
      "    \"category\": \"Mutation\"\n",
      "  },\n",
      "  {\n",
      "    \"entity\": \"A1762T, G1764A\",\n",
      "    \"confidence\": \"5\",\n",
      "    \"category\": \"Mutation\"\n",
      "  },\n",
      "  {\n",
      "    \"entity\": \"G1896A, G1899A\",\n",
      "    \"confidence\": \"5\",\n",
      "    \"category\": \"Mutation\"\n",
      "  }\n",
      "] [\n",
      "    {\n",
      "        \"entity\": \"Marfan syndrome (MFS)\",\n",
      "        \"confidence\": \"5\",\n",
      "        \"category\": \"Disease\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"fibrillin-1 gene (FBN1)\",\n",
      "        \"confidence\": \"5\",\n",
      "        \"category\": \"Gene\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"polymerase chain reaction (PCR)\",\n",
      "        \"confidence\": \"3\",\n",
      "        \"category\": \"MolecularFunction\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"Direct sequencing\",\n",
      "        \"confidence\": \"4\",\n",
      "        \"category\": \"Technique\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"Calcium binding epidermal growth factor-like#15 domain\",\n",
      "        \"confidence\": \"4\",\n",
      "        \"category\": \"ProteinDomain\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"Serine\",\n",
      "        \"confidence\": \"3\",\n",
      "        \"category\": \"AminoAcid\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"Proline\",\n",
      "        \"confidence\": \"3\",\n",
      "        \"category\": \"AminoAcid\"\n",
      "    }\n",
      "] [\n",
      "    {\n",
      "        \"entity\": \"Duchenne muscular dystrophy (DMD)\",\n",
      "        \"confidence\": \"5\",\n",
      "        \"category\": \"Disease\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"Becker muscular dystrophy (BMD)\",\n",
      "        \"confidence\": \"5\",\n",
      "        \"category\": \"Disease\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"DMD locus\",\n",
      "        \"confidence\": \"4\",\n",
      "        \"category\": \"Genome\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"dystrophin cDNA\",\n",
      "        \"confidence\": \"4\",\n",
      "        \"category\": \"Compound\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"cDNA5b-7\",\n",
      "        \"confidence\": \"4\",\n",
      "        \"category\": \"Compound\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"cDNA8\",\n",
      "        \"confidence\": \"4\",\n",
      "        \"category\": \"Compound\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"cDNA9\",\n",
      "        \"confidence\": \"4\",\n",
      "        \"category\": \"Compound\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"exons\",\n",
      "        \"confidence\": \"3\",\n",
      "        \"category\": \"BiologicalProcess\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"reading frame\",\n",
      "        \"confidence\": \"4\",\n",
      "        \"category\": \"MolecularFunction\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"DMD or BMD patients\",\n",
      "        \"confidence\": \"3\",\n",
      "        \"category\": \"Individual\"\n",
      "    }\n",
      "] [\n",
      "    {\n",
      "        \"entity\": \"Lithium\",\n",
      "        \"confidence\": \"5\",\n",
      "        \"category\": \"Compound\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"nephrogenic diabetes insipidus (NDI)\",\n",
      "        \"confidence\": \"5\",\n",
      "        \"category\": \"Disease\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"cAMP pathway\",\n",
      "        \"confidence\": \"4\",\n",
      "        \"category\": \"Pathway\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"PKC-mediated signaling\",\n",
      "        \"confidence\": \"4\",\n",
      "        \"category\": \"MolecularFunction\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"PKC-alpha (PKCa)\",\n",
      "        \"confidence\": \"4\",\n",
      "        \"category\": \"Protein\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"inner medulla\",\n",
      "        \"confidence\": \"3\",\n",
      "        \"category\": \"Anatomy\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"urea transporter (UT-A1)\",\n",
      "        \"confidence\": \"4\",\n",
      "        \"category\": \"Metabolite\"\n",
      "    },\n",
      "    {\n",
      "        \"entity\": \"water channel (AQP2)\",\n",
      "        \"confidence\": \"5\",\n",
      "        \"category\": \"Metabolite\"\n",
      "    }\n",
      "]"
     ]
    }
   ],
   "source": [
    "import os, sys\n",
    "import contextlib\n",
    "\n",
    "i = 0\n",
    "for data in datas['test']:\n",
    "    pmid = data['pmid']\n",
    "    passages = data['passages']\n",
    "    entities = data['entities']\n",
    "    abstract = passages[1]['text'][0]\n",
    "    abstract_offset = passages[1]['offsets'][0]\n",
    "    metadata = {\n",
    "        'pmid': pmid,\n",
    "        'type': 'abstract',\n",
    "    }\n",
    "\n",
    "    #with open(os.devnull, \"w\") as f, contextlib.redirect_stdout(f):\n",
    "    entities_extracted = extract_concepts(abstract, model=model_name, metadata=metadata)\n",
    "    \n",
    "    if entities_extracted:\n",
    "        disease_entities_extracted = [e for e in entities_extracted if e['category'] == 'Disease']\n",
    "        # print(disease_entities_extracted)\n",
    "        entities_str = json.dumps(entities_extracted, indent=4)\n",
    "        # print(entities_str)\n",
    "    else:\n",
    "        print(\"No entities found.\")\n",
    "    i += 1\n",
    "    if i >= 10:\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b8427f45-35f2-4ca4-9ca0-9c3f3eca7302",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'text': ['PKC-alpha'],\n",
       "  'offsets': [[11, 20]],\n",
       "  'concept_id': '18750',\n",
       "  'semantic_type_id': 'GeneOrGeneProduct'},\n",
       " {'text': ['lithium'],\n",
       "  'offsets': [[32, 39]],\n",
       "  'concept_id': 'D008094',\n",
       "  'semantic_type_id': 'ChemicalEntity'},\n",
       " {'text': ['nephrogenic diabetes insipidus'],\n",
       "  'offsets': [[48, 78]],\n",
       "  'concept_id': 'D018500',\n",
       "  'semantic_type_id': 'DiseaseOrPhenotypicFeature'},\n",
       " {'text': ['Lithium'],\n",
       "  'offsets': [[80, 87]],\n",
       "  'concept_id': 'D008094',\n",
       "  'semantic_type_id': 'ChemicalEntity'},\n",
       " {'text': ['nephrogenic diabetes insipidus'],\n",
       "  'offsets': [[125, 155]],\n",
       "  'concept_id': 'D018500',\n",
       "  'semantic_type_id': 'DiseaseOrPhenotypicFeature'},\n",
       " {'text': ['NDI'],\n",
       "  'offsets': [[157, 160]],\n",
       "  'concept_id': 'D018500',\n",
       "  'semantic_type_id': 'DiseaseOrPhenotypicFeature'},\n",
       " {'text': ['patients'],\n",
       "  'offsets': [[173, 181]],\n",
       "  'concept_id': '9606',\n",
       "  'semantic_type_id': 'OrganismTaxon'},\n",
       " {'text': ['lithium'],\n",
       "  'offsets': [[244, 251]],\n",
       "  'concept_id': 'D008094',\n",
       "  'semantic_type_id': 'ChemicalEntity'},\n",
       " {'text': ['cAMP'],\n",
       "  'offsets': [[275, 279]],\n",
       "  'concept_id': 'D000242',\n",
       "  'semantic_type_id': 'ChemicalEntity'},\n",
       " {'text': ['urea transporter'],\n",
       "  'offsets': [[313, 329]],\n",
       "  'concept_id': '27411',\n",
       "  'semantic_type_id': 'GeneOrGeneProduct'},\n",
       " {'text': ['UT-A1'],\n",
       "  'offsets': [[331, 336]],\n",
       "  'concept_id': '27411',\n",
       "  'semantic_type_id': 'GeneOrGeneProduct'},\n",
       " {'text': ['AQP2'],\n",
       "  'offsets': [[357, 361]],\n",
       "  'concept_id': '11827',\n",
       "  'semantic_type_id': 'GeneOrGeneProduct'},\n",
       " {'text': ['PKC'],\n",
       "  'offsets': [[448, 451]],\n",
       "  'concept_id': '18750',\n",
       "  'semantic_type_id': 'GeneOrGeneProduct'},\n",
       " {'text': ['lithium'],\n",
       "  'offsets': [[511, 518]],\n",
       "  'concept_id': 'D008094',\n",
       "  'semantic_type_id': 'ChemicalEntity'},\n",
       " {'text': ['polyuria'],\n",
       "  'offsets': [[527, 535]],\n",
       "  'concept_id': 'D011141',\n",
       "  'semantic_type_id': 'DiseaseOrPhenotypicFeature'},\n",
       " {'text': ['PKC-alpha'],\n",
       "  'offsets': [[537, 546]],\n",
       "  'concept_id': '18750',\n",
       "  'semantic_type_id': 'GeneOrGeneProduct'},\n",
       " {'text': ['mice'],\n",
       "  'offsets': [[552, 556]],\n",
       "  'concept_id': '10090',\n",
       "  'semantic_type_id': 'OrganismTaxon'},\n",
       " {'text': ['PKCa'],\n",
       "  'offsets': [[558, 562]],\n",
       "  'concept_id': '18750',\n",
       "  'semantic_type_id': 'GeneOrGeneProduct'},\n",
       " {'text': ['lithium'],\n",
       "  'offsets': [[628, 635]],\n",
       "  'concept_id': 'D008094',\n",
       "  'semantic_type_id': 'ChemicalEntity'},\n",
       " {'text': ['mice'],\n",
       "  'offsets': [[659, 663]],\n",
       "  'concept_id': '10090',\n",
       "  'semantic_type_id': 'OrganismTaxon'},\n",
       " {'text': ['PKCa'],\n",
       "  'offsets': [[760, 764]],\n",
       "  'concept_id': '18750',\n",
       "  'semantic_type_id': 'GeneOrGeneProduct'},\n",
       " {'text': ['mice'],\n",
       "  'offsets': [[768, 772]],\n",
       "  'concept_id': '10090',\n",
       "  'semantic_type_id': 'OrganismTaxon'},\n",
       " {'text': ['AQP2'],\n",
       "  'offsets': [[857, 861]],\n",
       "  'concept_id': '11827',\n",
       "  'semantic_type_id': 'GeneOrGeneProduct'},\n",
       " {'text': ['mice'],\n",
       "  'offsets': [[931, 935]],\n",
       "  'concept_id': '10090',\n",
       "  'semantic_type_id': 'OrganismTaxon'},\n",
       " {'text': ['AQP2'],\n",
       "  'offsets': [[946, 950]],\n",
       "  'concept_id': '11827',\n",
       "  'semantic_type_id': 'GeneOrGeneProduct'},\n",
       " {'text': ['PKCa'],\n",
       "  'offsets': [[968, 972]],\n",
       "  'concept_id': '18750',\n",
       "  'semantic_type_id': 'GeneOrGeneProduct'},\n",
       " {'text': ['UT-A1'],\n",
       "  'offsets': [[1012, 1017]],\n",
       "  'concept_id': '27411',\n",
       "  'semantic_type_id': 'GeneOrGeneProduct'},\n",
       " {'text': ['lithium'],\n",
       "  'offsets': [[1061, 1068]],\n",
       "  'concept_id': 'D008094',\n",
       "  'semantic_type_id': 'ChemicalEntity'},\n",
       " {'text': ['Lithium'],\n",
       "  'offsets': [[1082, 1089]],\n",
       "  'concept_id': 'D008094',\n",
       "  'semantic_type_id': 'ChemicalEntity'},\n",
       " {'text': ['mice'],\n",
       "  'offsets': [[1101, 1105]],\n",
       "  'concept_id': '10090',\n",
       "  'semantic_type_id': 'OrganismTaxon'},\n",
       " {'text': ['PKCa'],\n",
       "  'offsets': [[1157, 1161]],\n",
       "  'concept_id': '18750',\n",
       "  'semantic_type_id': 'GeneOrGeneProduct'},\n",
       " {'text': ['AQP2'],\n",
       "  'offsets': [[1206, 1210]],\n",
       "  'concept_id': '11827',\n",
       "  'semantic_type_id': 'GeneOrGeneProduct'},\n",
       " {'text': ['UT-A1'],\n",
       "  'offsets': [[1215, 1220]],\n",
       "  'concept_id': '27411',\n",
       "  'semantic_type_id': 'GeneOrGeneProduct'},\n",
       " {'text': ['lithium'],\n",
       "  'offsets': [[1254, 1261]],\n",
       "  'concept_id': 'D008094',\n",
       "  'semantic_type_id': 'ChemicalEntity'},\n",
       " {'text': ['PKCa'],\n",
       "  'offsets': [[1300, 1304]],\n",
       "  'concept_id': '18750',\n",
       "  'semantic_type_id': 'GeneOrGeneProduct'},\n",
       " {'text': ['mice'],\n",
       "  'offsets': [[1308, 1312]],\n",
       "  'concept_id': '10090',\n",
       "  'semantic_type_id': 'OrganismTaxon'},\n",
       " {'text': ['AQP2'],\n",
       "  'offsets': [[1314, 1318]],\n",
       "  'concept_id': '11827',\n",
       "  'semantic_type_id': 'GeneOrGeneProduct'},\n",
       " {'text': ['UT-A1'],\n",
       "  'offsets': [[1350, 1355]],\n",
       "  'concept_id': '27411',\n",
       "  'semantic_type_id': 'GeneOrGeneProduct'},\n",
       " {'text': ['sodium'],\n",
       "  'offsets': [[1391, 1397]],\n",
       "  'concept_id': 'D012964',\n",
       "  'semantic_type_id': 'ChemicalEntity'},\n",
       " {'text': ['potassium'],\n",
       "  'offsets': [[1399, 1408]],\n",
       "  'concept_id': 'D011188',\n",
       "  'semantic_type_id': 'ChemicalEntity'},\n",
       " {'text': ['calcium'],\n",
       "  'offsets': [[1413, 1420]],\n",
       "  'concept_id': 'D002118',\n",
       "  'semantic_type_id': 'ChemicalEntity'},\n",
       " {'text': ['lithium'],\n",
       "  'offsets': [[1438, 1445]],\n",
       "  'concept_id': 'D008094',\n",
       "  'semantic_type_id': 'ChemicalEntity'},\n",
       " {'text': ['lithium'],\n",
       "  'offsets': [[1464, 1471]],\n",
       "  'concept_id': 'D008094',\n",
       "  'semantic_type_id': 'ChemicalEntity'},\n",
       " {'text': ['PKCa'],\n",
       "  'offsets': [[1476, 1480]],\n",
       "  'concept_id': '18750',\n",
       "  'semantic_type_id': 'GeneOrGeneProduct'},\n",
       " {'text': ['mice'],\n",
       "  'offsets': [[1484, 1488]],\n",
       "  'concept_id': '10090',\n",
       "  'semantic_type_id': 'OrganismTaxon'},\n",
       " {'text': ['PKCa'],\n",
       "  'offsets': [[1521, 1525]],\n",
       "  'concept_id': '18750',\n",
       "  'semantic_type_id': 'GeneOrGeneProduct'},\n",
       " {'text': ['AQP2'],\n",
       "  'offsets': [[1536, 1540]],\n",
       "  'concept_id': '11827',\n",
       "  'semantic_type_id': 'GeneOrGeneProduct'},\n",
       " {'text': ['UT-A1'],\n",
       "  'offsets': [[1545, 1550]],\n",
       "  'concept_id': '27411',\n",
       "  'semantic_type_id': 'GeneOrGeneProduct'},\n",
       " {'text': ['lithium'],\n",
       "  'offsets': [[1590, 1597]],\n",
       "  'concept_id': 'D008094',\n",
       "  'semantic_type_id': 'ChemicalEntity'},\n",
       " {'text': ['NDI'],\n",
       "  'offsets': [[1606, 1609]],\n",
       "  'concept_id': 'D018500',\n",
       "  'semantic_type_id': 'DiseaseOrPhenotypicFeature'},\n",
       " {'text': ['polyuria'],\n",
       "  'offsets': [[1654, 1662]],\n",
       "  'concept_id': 'D011141',\n",
       "  'semantic_type_id': 'DiseaseOrPhenotypicFeature'},\n",
       " {'text': ['lithium'],\n",
       "  'offsets': [[1679, 1686]],\n",
       "  'concept_id': 'D008094',\n",
       "  'semantic_type_id': 'ChemicalEntity'}]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "entities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "dadd8a91-44b9-427a-8a4e-63b5b3f59d3c",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'entity': 'Lithium',\n",
       "  'confidence': '5',\n",
       "  'category': 'Compound',\n",
       "  'pmid': '25006961',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'nephrogenic diabetes insipidus (NDI)',\n",
       "  'confidence': '5',\n",
       "  'category': 'Disease',\n",
       "  'pmid': '25006961',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'cAMP pathway',\n",
       "  'confidence': '4',\n",
       "  'category': 'Pathway',\n",
       "  'pmid': '25006961',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'PKC-mediated signaling',\n",
       "  'confidence': '4',\n",
       "  'category': 'MolecularFunction',\n",
       "  'pmid': '25006961',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'PKC-alpha (PKCa)',\n",
       "  'confidence': '4',\n",
       "  'category': 'Protein',\n",
       "  'pmid': '25006961',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'inner medulla',\n",
       "  'confidence': '3',\n",
       "  'category': 'Anatomy',\n",
       "  'pmid': '25006961',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'urea transporter (UT-A1)',\n",
       "  'confidence': '4',\n",
       "  'category': 'Metabolite',\n",
       "  'pmid': '25006961',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'water channel (AQP2)',\n",
       "  'confidence': '5',\n",
       "  'category': 'Metabolite',\n",
       "  'pmid': '25006961',\n",
       "  'type': 'abstract'}]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "entities_extracted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "182962b2-3859-49aa-bf64-0bdc8baba8d2",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'entity': 'nephrogenic diabetes insipidus (NDI)',\n",
       "  'confidence': '5',\n",
       "  'category': 'Disease',\n",
       "  'pmid': '25006961',\n",
       "  'type': 'abstract'}]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "disease_entities_extracted = [e for e in entities_extracted if e['category'] == 'Disease']\n",
    "disease_entities_extracted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "088cdf0c-0386-4278-9a50-aa78eaceaf23",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b26dc6db-729c-4f23-822b-bed2b7671cd8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd5e9502-f7e8-4fe7-a500-46666bd9aa8e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d1b8997e-4bfa-4180-a27d-245ed4ef0493",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[122, 1815]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "abstract_offset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "936e8d88-21be-465d-b483-8202cd5dd037",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'OBJECTIVE: Congenital long QT syndrome (LQTS) with in utero onset of the rhythm disturbances is associated with a poor prognosis. In this study we investigated a newborn patient with fetal bradycardia, 2:1 atrioventricular block and ventricular tachycardia soon after birth. METHODS: Mutational analysis and DNA sequencing were conducted in a newborn. The 2:1 atrioventricular block improved to 1:1 conduction only after intravenous lidocaine infusion or a high dose of mexiletine, which also controlled the ventricular tachycardia. RESULTS: A novel, spontaneous LQTS-3 mutation was identified in the transmembrane segment 6 of domain IV of the Na(v)1.5 cardiac sodium channel, with a G-->A substitution at codon 1763, which changed a valine (GTG) to a methionine (ATG). The proband was heterozygous but the mutation was absent in the parents and the sister. Expression of this mutant channel in tsA201 mammalian cells by site-directed mutagenesis revealed a persistent tetrodotoxin-sensitive but lidocaine-resistant current that was associated with a positive shift of the steady-state inactivation curve, steeper activation curve and faster recovery from inactivation. We also found a similar electrophysiological profile for the neighboring V1764M mutant. But, the other neighboring I1762A mutant had no persistent current and was still associated with a positive shift of inactivation. CONCLUSIONS: These findings suggest that the Na(v)1.5/V1763M channel dysfunction and possible neighboring mutants contribute to a persistent inward current due to altered inactivation kinetics and clinically congenital LQTS with perinatal onset of arrhythmias that responded to lidocaine and mexiletine.'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "abstract"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b9abd114-9df0-4128-8f64-de31bdd3bd8f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def highlight_entities_label(s, start_idx, entities, color='#66CC66'):\n",
    "    left = f\"<span style='background-color: {color}'>\"\n",
    "    right = \"</span>\"\n",
    "    final = \"\"\n",
    "    last_r = 0\n",
    "    for entity in entities:\n",
    "        l, r = entity['offsets'][0]\n",
    "        if not l >= start_idx:\n",
    "            continue\n",
    "        l -= start_idx\n",
    "        r -= start_idx\n",
    "        final += s[last_r:l]\n",
    "        final += left + s[l:r] + right\n",
    "        last_r = r\n",
    "    final += s[r:]\n",
    "    return final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "9e348e99-504f-42db-9488-dd0c6223843e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "OBJECTIVE: Congenital <span style='background-color: #66CC66'>long QT syndrome</span> (<span style='background-color: #66CC66'>LQTS</span>) with in utero onset of the rhythm disturbances is associated with a poor prognosis. In this study we investigated a newborn <span style='background-color: #66CC66'>patient</span> with fetal <span style='background-color: #66CC66'>bradycardia</span>, 2:1 <span style='background-color: #66CC66'>atrioventricular block</span> and <span style='background-color: #66CC66'>ventricular tachycardia</span> soon after birth. METHODS: Mutational analysis and DNA sequencing were conducted in a newborn. The 2:1 <span style='background-color: #66CC66'>atrioventricular block</span> improved to 1:1 conduction only after intravenous <span style='background-color: #66CC66'>lidocaine</span> infusion or a high dose of <span style='background-color: #66CC66'>mexiletine</span>, which also controlled the <span style='background-color: #66CC66'>ventricular tachycardia</span>. RESULTS: A novel, spontaneous <span style='background-color: #66CC66'>LQTS</span>-3 mutation was identified in the transmembrane segment 6 of domain IV of the <span style='background-color: #66CC66'>Na(v)1.5</span> cardiac <span style='background-color: #66CC66'>sodium</span> channel, with a <span style='background-color: #66CC66'>G-->A substitution at codon 1763</span>, which changed a <span style='background-color: #66CC66'>valine (GTG) to a methionine (ATG)</span>. The proband was heterozygous but the mutation was absent in the parents and the sister. Expression of this mutant channel in <span style='background-color: #66CC66'>tsA201</span> mammalian cells by site-directed mutagenesis revealed a persistent <span style='background-color: #66CC66'>tetrodotoxin</span>-sensitive but <span style='background-color: #66CC66'>lidocaine</span>-resistant current that was associated with a positive shift of the steady-state inactivation curve, steeper activation curve and faster recovery from inactivation. We also found a similar electrophysiological profile for the neighboring <span style='background-color: #66CC66'>V1764M</span> mutant. But, the other neighboring <span style='background-color: #66CC66'>I1762A</span> mutant had no persistent current and was still associated with a positive shift of inactivation. CONCLUSIONS: These findings suggest that the <span style='background-color: #66CC66'>Na(v)1.5</span>/<span style='background-color: #66CC66'>V1763M</span> channel dysfunction and possible neighboring mutants contribute to a persistent inward current due to altered inactivation kinetics and clinically congenital <span style='background-color: #66CC66'>LQTS</span> with perinatal onset of <span style='background-color: #66CC66'>arrhythmias</span> that responded to <span style='background-color: #66CC66'>lidocaine</span> and <span style='background-color: #66CC66'>mexiletine</span>."
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import HTML\n",
    "\n",
    "label = highlight_entities_label(abstract, abstract_offset[0], entities)\n",
    "HTML(label)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ac1387a0-dd26-42bb-8c66-92917cf93f9f",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "OBJECTIVE: <span style='background-color: #00CCFF'>Congenital long QT syndrome (LQTS)</span> with in utero onset of the rhythm disturbances is associated with a poor prognosis. In this study we investigated a newborn patient with fetal bradycardia, 2:1 atrioventricular block and ventricular tachycardia soon after birth. METHODS: Mutational analysis and DNA sequencing were conducted in a newborn. The 2:1 atrioventricular block improved to 1:1 conduction only after intravenous <span style='background-color: #00CCFF'>Lidocaine</span> infusion or a high dose of <span style='background-color: #00CCFF'>Mexiletine</span>, which also controlled the ventricular tachycardia. RESULTS: A novel, spontaneous <span style='background-color: #00CCFF'>LQTS-3</span> mutation was identified in the <span style='background-color: #00CCFF'>Transmembrane segment 6 of domain IV</span> of the <span style='background-color: #00CCFF'>Na(v)1.5 cardiac sodium channel</span>, with a <span style='background-color: #00CCFF'>G-->A substitution at codon 1763</span>, which changed a <span style='background-color: #00CCFF'>Valine (GTG)</span> to a <span style='background-color: #00CCFF'>Methionine (ATG)</span>. The proband was heterozygous but the mutation was absent in the parents and the sister. Expression of this mutant channel in <span style='background-color: #00CCFF'>tsA201 mammalian cells</span> by <span style='background-color: #00CCFF'>Site-directed mutagenesis</span> revealed a persistent tetrodotoxin-sensitive but <span style='background-color: #00CCFF'>Lidocaine</span>-resistant current that was associated with a positive shift of the steady-state inactivation curve, steeper activation curve and faster recovery from inactivation. We also found a similar electrophysiological profile for the neighboring <span style='background-color: #00CCFF'>V1764M mutant</span>. But, the other neighboring <span style='background-color: #00CCFF'>I1762A mutant</span> had no persistent current and was still associated with a positive shift of inactivation. CONCLUSIONS: These findings suggest that the Na(v)1.5/V1763M channel dysfunction and possible neighboring mutants contribute to a persistent inward current due to altered inactivation kinetics and clinically congenital LQTS with perinatal onset of arrhythmias that responded to <span style='background-color: #00CCFF'>Lidocaine</span> and <span style='background-color: #00CCFF'>Mexiletine</span>."
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def highlight_entities_extracted(s, entities, color='#00CCFF'):\n",
    "    import re\n",
    "    left = f\"<span style='background-color: {color}'>\"\n",
    "    right = \"</span>\"\n",
    "    final = s[:]\n",
    "    for entity in entities:\n",
    "        entity_text = entity['entity']\n",
    "        final = re.sub('(?i)'+re.escape(entity_text), lambda m: left+entity_text+right, final)\n",
    "    return final\n",
    "predicted = highlight_entities_extracted(abstract, entities_extracted)\n",
    "HTML(predicted)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "dc04aadb-db35-4193-8dc7-4e72bbce8e3f",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'entity': 'Congenital long QT syndrome (LQTS)',\n",
       "  'importance': 5,\n",
       "  'category': 'Disease',\n",
       "  'pmid': '15485686',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'In utero onset of rhythm disturbances',\n",
       "  'importance': 3,\n",
       "  'category': 'BiologicalProcess',\n",
       "  'pmid': '15485686',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'Newborn patient',\n",
       "  'importance': 4,\n",
       "  'category': 'Anatomy',\n",
       "  'pmid': '15485686',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'Fetal bradycardia',\n",
       "  'importance': 3,\n",
       "  'category': 'Symptom',\n",
       "  'pmid': '15485686',\n",
       "  'type': 'abstract'},\n",
       " {'entity': '2:1 atrioventricular block',\n",
       "  'importance': 4,\n",
       "  'category': 'Pathway',\n",
       "  'pmid': '15485686',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'Ventricular tachycardia',\n",
       "  'importance': 3,\n",
       "  'category': 'Symptom',\n",
       "  'pmid': '15485686',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'Mutational analysis and DNA sequencing',\n",
       "  'importance': 4,\n",
       "  'category': 'MolecularFunction',\n",
       "  'pmid': '15485686',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'Lidocaine infusion',\n",
       "  'importance': 3,\n",
       "  'category': 'Compound',\n",
       "  'pmid': '15485686',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'High dose of mexiletine',\n",
       "  'importance': 3,\n",
       "  'category': 'Compound',\n",
       "  'pmid': '15485686',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'Na(v)1.5 cardiac sodium channel',\n",
       "  'importance': 5,\n",
       "  'category': 'Anatomy',\n",
       "  'pmid': '15485686',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'Transmembrane segment 6 of domain IV',\n",
       "  'importance': 3,\n",
       "  'category': 'CellularComponent',\n",
       "  'pmid': '15485686',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'G-->A substitution at codon 1763',\n",
       "  'importance': 4,\n",
       "  'category': 'MolecularFunction',\n",
       "  'pmid': '15485686',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'Valine (GTG)',\n",
       "  'importance': 2,\n",
       "  'category': 'Compound',\n",
       "  'pmid': '15485686',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'Methionine (ATG)',\n",
       "  'importance': 2,\n",
       "  'category': 'Compound',\n",
       "  'pmid': '15485686',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'Persistent tetrodotoxin-sensitive but lidocaine-resistant current',\n",
       "  'importance': 4,\n",
       "  'category': 'Metabolite',\n",
       "  'pmid': '15485686',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'Positive shift of the steady-state inactivation curve',\n",
       "  'importance': 3,\n",
       "  'category': 'BiologicalProcess',\n",
       "  'pmid': '15485686',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'Steep activation curve',\n",
       "  'importance': 2,\n",
       "  'category': 'BiologicalProcess',\n",
       "  'pmid': '15485686',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'Faster recovery from inactivation',\n",
       "  'importance': 3,\n",
       "  'category': 'BiologicalProcess',\n",
       "  'pmid': '15485686',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'V1764M mutant',\n",
       "  'importance': 3,\n",
       "  'category': 'Mutation',\n",
       "  'pmid': '15485686',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'I1762A mutant',\n",
       "  'importance': 2,\n",
       "  'category': 'Mutation',\n",
       "  'pmid': '15485686',\n",
       "  'type': 'abstract'}]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "entities_extracted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "1bf78dc6-55dc-4842-a706-171e672b201c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "1cfc35ac-d352-4822-b83c-fe663ccd0c43",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "fcaf194f-0ef0-4075-9c93-4a5b160d4109",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'bye bye bye'"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "re.sub('hello', 'bye', 'hello HeLLo HELLO', flags=re.IGNORECASE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "c9c020a2-75da-42df-a7dd-6315cb08b057",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Congenital long QT syndrome (LQTS)'"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "re.sub('Congenital long QT syndrome (LQTS)', 'hello', 'Congenital long QT syndrome (LQTS)', flags=re.IGNORECASE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "d26199a1-9207-493b-8334-212378e9a2de",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "confusion_matrix = {\"test\":[0,0,0]}\n",
    "categories = []\n",
    "\n",
    "def match_entity(entity_a, entity_b):\n",
    "    return entity_a['category'] == entity_b['category'] and\\\n",
    "model.query(entity_a['entity'], top_k=1)[2] == model.query(entity_b['entity'], top_k=1)[2]\n",
    "\n",
    "def macro_avg_f1_score(ans, gt):\n",
    "    # for sample_ans, sample_gt in zip(ans, gt):\n",
    "        # for category in categories:\n",
    "            # get all entities in the category\n",
    "    # ans_entities = [entity['entity']\n",
    "    #                 for entity in sample_ans if entity['category'] == category]\n",
    "    # ans_entities = [entity['entity'] for entity in ans]\n",
    "    # # gt_entities = [entity['text']\n",
    "    # #                for entity in sample_gt if entity['category'] == category]\n",
    "    # gt_entities = [entity['text'][0] for entity in gt]\n",
    "    # TP = len(set(ans_entities) & set(gt_entities))\n",
    "    # FP = len(set(ans_entities) - set(gt_entities))\n",
    "    # FN = len(set(gt_entities) - set(ans_entities))\n",
    "    # confusion_matrix[\"test\"][0] += TP\n",
    "    # confusion_matrix[\"test\"][1] += FP\n",
    "    # confusion_matrix[\"test\"][2] += FN\n",
    "    # f1_score = []\n",
    "    # # for category in categories:\n",
    "    # TP, FP, FN = confusion_matrix[\"test\"]\n",
    "    # precision = TP/(TP+FP) if TP+FP else float('nan')\n",
    "    # recall = TP/(TP+FN) if TP+FN else float('nan')\n",
    "    # f1_score.append(2*(precision * recall) / (precision + recall))\n",
    "    # return sum(f1_score) / len(f1_score)\n",
    "\n",
    "macro_avg_f1_score(entities_extracted, entities)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "4faa1853-24fe-4d37-8d31-da9fa1f8c1da",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'entity': 'Congenital long QT syndrome (LQTS)',\n",
       "  'importance': 5,\n",
       "  'category': 'Disease',\n",
       "  'pmid': '15485686',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'In utero onset of rhythm disturbances',\n",
       "  'importance': 3,\n",
       "  'category': 'BiologicalProcess',\n",
       "  'pmid': '15485686',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'Newborn patient',\n",
       "  'importance': 4,\n",
       "  'category': 'Anatomy',\n",
       "  'pmid': '15485686',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'Fetal bradycardia',\n",
       "  'importance': 3,\n",
       "  'category': 'Symptom',\n",
       "  'pmid': '15485686',\n",
       "  'type': 'abstract'},\n",
       " {'entity': '2:1 atrioventricular block',\n",
       "  'importance': 4,\n",
       "  'category': 'Pathway',\n",
       "  'pmid': '15485686',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'Ventricular tachycardia',\n",
       "  'importance': 3,\n",
       "  'category': 'Symptom',\n",
       "  'pmid': '15485686',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'Mutational analysis and DNA sequencing',\n",
       "  'importance': 4,\n",
       "  'category': 'MolecularFunction',\n",
       "  'pmid': '15485686',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'Lidocaine infusion',\n",
       "  'importance': 3,\n",
       "  'category': 'Compound',\n",
       "  'pmid': '15485686',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'High dose of mexiletine',\n",
       "  'importance': 3,\n",
       "  'category': 'Compound',\n",
       "  'pmid': '15485686',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'Na(v)1.5 cardiac sodium channel',\n",
       "  'importance': 5,\n",
       "  'category': 'Anatomy',\n",
       "  'pmid': '15485686',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'Transmembrane segment 6 of domain IV',\n",
       "  'importance': 3,\n",
       "  'category': 'CellularComponent',\n",
       "  'pmid': '15485686',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'G-->A substitution at codon 1763',\n",
       "  'importance': 4,\n",
       "  'category': 'MolecularFunction',\n",
       "  'pmid': '15485686',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'Valine (GTG)',\n",
       "  'importance': 2,\n",
       "  'category': 'Compound',\n",
       "  'pmid': '15485686',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'Methionine (ATG)',\n",
       "  'importance': 2,\n",
       "  'category': 'Compound',\n",
       "  'pmid': '15485686',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'Persistent tetrodotoxin-sensitive but lidocaine-resistant current',\n",
       "  'importance': 4,\n",
       "  'category': 'Metabolite',\n",
       "  'pmid': '15485686',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'Positive shift of the steady-state inactivation curve',\n",
       "  'importance': 3,\n",
       "  'category': 'BiologicalProcess',\n",
       "  'pmid': '15485686',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'Steep activation curve',\n",
       "  'importance': 2,\n",
       "  'category': 'BiologicalProcess',\n",
       "  'pmid': '15485686',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'Faster recovery from inactivation',\n",
       "  'importance': 3,\n",
       "  'category': 'BiologicalProcess',\n",
       "  'pmid': '15485686',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'V1764M mutant',\n",
       "  'importance': 3,\n",
       "  'category': 'Mutation',\n",
       "  'pmid': '15485686',\n",
       "  'type': 'abstract'},\n",
       " {'entity': 'I1762A mutant',\n",
       "  'importance': 2,\n",
       "  'category': 'Mutation',\n",
       "  'pmid': '15485686',\n",
       "  'type': 'abstract'}]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "entities_extracted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b1edc533-5c0f-4760-8b88-84e142aff668",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'text': ['long QT syndrome'],\n",
       " 'offsets': [[56, 72]],\n",
       " 'concept_id': 'D008133',\n",
       " 'semantic_type_id': 'DiseaseOrPhenotypicFeature'}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "entities[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "29885e07-3b4d-4be2-ac6f-50d15fb1cdaa",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'entity': 'Congenital long QT syndrome (LQTS)',\n",
       " 'importance': 5,\n",
       " 'category': 'Disease',\n",
       " 'pmid': '15485686',\n",
       " 'type': 'abstract'}"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "entities_extracted[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "658036e5-52f5-4f40-9875-ec8bd370a1fc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'CellLine',\n",
       " 'ChemicalEntity',\n",
       " 'DiseaseOrPhenotypicFeature',\n",
       " 'GeneOrGeneProduct',\n",
       " 'OrganismTaxon',\n",
       " 'SequenceVariant'}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set([j['semantic_type_id'] for i in datas[\"test\"] for j in i['entities']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17364ae0-4324-49c1-9b6a-c6933505c04a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from text2knowledge.normalizer import Normalizer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f04243a3-f882-4e15-80cd-00c3e9ece689",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "class Arg:\n",
    "    def __init__(self):\n",
    "        # self.mention = \"G-->A substitution at codon 1763\"\n",
    "        self.model_name_or_path = \"dmis-lab/biosyn-biobert-bc2gn\"\n",
    "        self.show_embeddings = False\n",
    "        self.show_predictions = True\n",
    "        self.dictionary_path = \"/root/text2knowledge/data/dictionary/dict_Gene.txt\"\n",
    "        self.use_cuda = True\n",
    "arg = Arg()\n",
    "normalizer = Normalizer(arg)\n",
    "normalizer.normalize(\"G-->A substitution at codon 1763\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b615c9cb-0177-47bb-a7f7-60a2cacec8fb",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "'NoneType' object is not callable",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[7], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mnormalizer\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbiosyn\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43membed_sparse\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnames\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mG-->A substitution at codon 1763\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.8/site-packages/biosyn-0.1-py3.8.egg/biosyn/biosyn.py:209\u001b[0m, in \u001b[0;36mBioSyn.embed_sparse\u001b[0;34m(self, names, show_progress)\u001b[0m\n\u001b[1;32m    207\u001b[0m end \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mmin\u001b[39m(start \u001b[38;5;241m+\u001b[39m batch_size, \u001b[38;5;28mlen\u001b[39m(names))\n\u001b[1;32m    208\u001b[0m batch \u001b[38;5;241m=\u001b[39m names[start:end]\n\u001b[0;32m--> 209\u001b[0m batch_sparse_embeds \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msparse_encoder\u001b[49m\u001b[43m(\u001b[49m\u001b[43mbatch\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    210\u001b[0m batch_sparse_embeds \u001b[38;5;241m=\u001b[39m batch_sparse_embeds\u001b[38;5;241m.\u001b[39mnumpy()\n\u001b[1;32m    211\u001b[0m sparse_embeds\u001b[38;5;241m.\u001b[39mappend(batch_sparse_embeds)\n",
      "\u001b[0;31mTypeError\u001b[0m: 'NoneType' object is not callable"
     ]
    }
   ],
   "source": [
    "normalizer.biosyn.embed_sparse(names=[\"G-->A substitution at codon 1763\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5245f69d-3a74-46dd-8898-48cc85be1719",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "KeyboardInterrupt\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from text2knowledge.normalizer import normalize\n",
    "\n",
    "class Arg:\n",
    "    def __init__(self):\n",
    "        self.mention = \"long QT syndrome (LQTS)\"\n",
    "        self.model_name_or_path = \"dmis-lab/biosyn-biobert-ncbi-disease\"\n",
    "        self.show_embeddings = False\n",
    "        self.show_predictions = True\n",
    "        self.dictionary_path = \"/root/text2knowledge/data/dictionary/dict_Disease_20210630.txt\"\n",
    "        self.use_cuda = True\n",
    "arg = Arg()\n",
    "normalize(arg)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ba1393ce-3242-42a4-ad02-8f4cd40703fb",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from text2knowledge.normalizer import Normalizer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "fa47c888-e7a8-4229-829b-214636eb7e42",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/root/miniconda3/lib/python3.8/site-packages/huggingface_hub/file_download.py:678: FutureWarning: 'cached_download' is the legacy way to download files from the HF hub, please consider upgrading to 'hf_hub_download'\n",
      "  warnings.warn(\n",
      "/root/miniconda3/lib/python3.8/site-packages/sklearn/base.py:348: InconsistentVersionWarning: Trying to unpickle estimator TfidfTransformer from version 0.21.2 when using version 1.3.2. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:\n",
      "https://scikit-learn.org/stable/model_persistence.html#security-maintainability-limitations\n",
      "  warnings.warn(\n",
      "/root/miniconda3/lib/python3.8/site-packages/sklearn/base.py:348: InconsistentVersionWarning: Trying to unpickle estimator TfidfVectorizer from version 0.21.2 when using version 1.3.2. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:\n",
      "https://scikit-learn.org/stable/model_persistence.html#security-maintainability-limitations\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "class Arg:\n",
    "    def __init__(self):\n",
    "        # self.mention = \"long QT syndrome (LQTS)\"\n",
    "        self.model_name_or_path = \"dmis-lab/biosyn-biobert-ncbi-disease\"\n",
    "        self.show_embeddings = False\n",
    "        self.show_predictions = True\n",
    "        self.dictionary_path = \"/root/text2knowledge/data/dictionary/dict_Disease_20210630.txt\"\n",
    "        self.use_cuda = True\n",
    "arg = Arg()\n",
    "\n",
    "disease_normalizer = Normalizer(arg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6b57b766-6b76-46c4-aec4-d47c8c26f1f9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/root/miniconda3/lib/python3.8/site-packages/biosyn-0.1-py3.8.egg/biosyn/biosyn.py:26: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded dictionary from cached file ./tmp/cached_biosyn-biobert-ncbi-disease_dict_Disease_20210630.pk\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'mention': 'long qt syndrome lqts',\n",
       " 'predictions': [{'name': 'Long QT Syndrome', 'id': 'MESH:D008133'},\n",
       "  {'name': 'Long Qt Syndrome 1/2', 'id': 'MESH:C566006'},\n",
       "  {'name': 'Long Qt Syndrome 2|LONG QT SYNDROME 1/2, DIGENIC, INCLUDED;LQT1/2, DIGENIC, INCLUDED|LONG QT SYNDROME 2/3, DIGENIC, INCLUDED;LQT2/3, DIGENIC, INCLUDED|LONG QT SYNDROME 2/5, DIGENIC, INCLUDED;LQT2/5, DIGENIC, INCLUDED|LONG QT SYNDROME 2/9, DIGENIC, INCLUDED;LQT2/9, DIGENIC, INCLUDED|LONG QT SYNDROME 2, ACQUIRED, SUSCEPTIBILITY TO, INCLUDED|LQT2',\n",
       "   'id': 'MESH:C563614'},\n",
       "  {'name': 'Long Qt Syndrome 10', 'id': 'OMIM:611819,MESH:C567514'},\n",
       "  {'name': 'LONG QT SYNDROME 13', 'id': 'OMIM:613485'}]}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "disease_normalizer.normalize(\"long QT syndrome (LQTS)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "58c1d39f-bd1f-4d69-8e18-c0f22a22a44f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded dictionary from cached file ./tmp/cached_biosyn-biobert-ncbi-disease_dict_Disease_20210630.pk\n",
      "{'mention': 'nephrogenic diabetes insipidus ndi',\n",
      " 'predictions': [{'name': 'diabetes insipidus, nephrogenic, x-linked|congenital nephrogenic diabetes insipidus|diabetes insipidus, nephrogenic, type ii|vasopressin-resistant diabetes insipidus|nephrogenic diabetes insipidus, type is|diabetes insipidus, nephrogenic, autosomal|acquired nephrogenic diabetes insipiduses|congenital nephrogenic diabetes insipiduses|acquired nephrogenic diabetes insipidus|diabetes insipidus renalis|adh-resistant diabetes insipidus|nephrogenic diabetes insipidus|adh-resistant diabetes insipiduses|diabetes insipidus, nephrogenics|diabetes insipidus, nephrogenic, type 1|diabetes insipidus, nephrogenic|diabetes insipidus, nephrogenic, type i|nephrogenic diabetes insipidus, type iis|nephrogenic diabetes insipidus, type i|nephrogenic diabetes insipidus, type ii|vasopressin resistant diabetes insipidus|Familial nephrogenic diabetes insipidus|Congenital nephrogenic diabetes insipidus|ADH-resistant diabetes insipidus|Renal diabetes insipidus|Hereditary nephrogenic diabetes insipidus|Vasopressin-resistant diabetes insipidus|Vasopressin-resistant hyposthenuria|Diabetes Insipidus, Nephrogenic|Nephrogenic diabetes insipidus|NDI - Nephrogenic diabetes insipidus|DIABETES INSIPIDUS NEPHROGENIC|Acquired nephrogenic diabetes insipidus|-Acquired Nephrogenic Diabetes Insipidus|Secondary nephrogenic diabetes insipidus|Acquired Nephrogenic Diabetes Insipidus|ADH-Resistant Diabetes Insipidus|Congenital Nephrogenic Diabetes Insipidus|Diabetes Insipidus, Nephrogenic, Autosomal|Diabetes Insipidus, Nephrogenic, Type 1|Diabetes Insipidus, Nephrogenic, Type I|Diabetes Insipidus, Nephrogenic, Type II|Diabetes Insipidus, Nephrogenic, X-Linked|Diabetes Insipidus Renalis|Nephrogenic Diabetes Insipidus|Nephrogenic Diabetes Insipidus, Type I|Nephrogenic Diabetes Insipidus, Type II|Vasopressin-Resistant Diabetes Insipidus',\n",
      "                  'id': 'OMIM:125800,MESH:D018500'},\n",
      "                 {'name': 'Dipsogenic Diabetes Insipidus',\n",
      "                  'id': 'MESH:C548013'},\n",
      "                 {'name': 'Diabetes Mellitus, Noninsulin-Dependent, 1|Noninsulin-Dependent Diabetes Mellitus 1',\n",
      "                  'id': 'OMIM:601283,MESH:C563359'},\n",
      "                 {'name': 'diabetes insipiduses|diabetes insipidus|DI - Diabetes insipidus|Diabetes insipidus|Diabetes insipidus (disorder)|Diabetes Insipidus',\n",
      "                  'id': 'MESH:D003919'},\n",
      "                 {'name': 'Nephrogenic Syndrome of Inappropriate Antidiuresis|NSIAD',\n",
      "                  'id': 'OMIM:300539,MESH:C564491'}]}\n"
     ]
    }
   ],
   "source": [
    "import pprint\n",
    "for disease_entity in disease_entities_extracted:\n",
    "    result = disease_normalizer.normalize(disease_entity[\"entity\"])\n",
    "    pprint.pp(result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c92943f7-fd60-48c1-8eb2-9df3792f7f45",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'entities_extracted' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[6], line 33\u001b[0m\n\u001b[1;32m      9\u001b[0m     \u001b[38;5;28;01mpass\u001b[39;00m\n\u001b[1;32m     10\u001b[0m     \u001b[38;5;66;03m# for sample_ans, sample_gt in zip(ans, gt):\u001b[39;00m\n\u001b[1;32m     11\u001b[0m         \u001b[38;5;66;03m# for category in categories:\u001b[39;00m\n\u001b[1;32m     12\u001b[0m             \u001b[38;5;66;03m# get all entities in the category\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     30\u001b[0m     \u001b[38;5;66;03m# f1_score.append(2*(precision * recall) / (precision + recall))\u001b[39;00m\n\u001b[1;32m     31\u001b[0m     \u001b[38;5;66;03m# return sum(f1_score) / len(f1_score)\u001b[39;00m\n\u001b[0;32m---> 33\u001b[0m macro_avg_f1_score(\u001b[43mentities_extracted\u001b[49m, entities)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'entities_extracted' is not defined"
     ]
    }
   ],
   "source": [
    "confusion_matrix = {\"test\":[0,0,0]}\n",
    "categories = []\n",
    "\n",
    "def match_entity(entity_a, entity_b):\n",
    "    return entity_a['category'] == entity_b['category'] and\\\n",
    "model.query(entity_a['entity'], top_k=1)[2] == model.query(entity_b['entity'], top_k=1)[2]\n",
    "\n",
    "def macro_avg_f1_score(ans, gt):\n",
    "    pass\n",
    "    # for sample_ans, sample_gt in zip(ans, gt):\n",
    "        # for category in categories:\n",
    "            # get all entities in the category\n",
    "    # ans_entities = [entity['entity']\n",
    "    #                 for entity in sample_ans if entity['category'] == category]\n",
    "    # ans_entities = [entity['entity'] for entity in ans]\n",
    "    # # gt_entities = [entity['text']\n",
    "    # #                for entity in sample_gt if entity['category'] == category]\n",
    "    # gt_entities = [entity['text'][0] for entity in gt]\n",
    "    # TP = len(set(ans_entities) & set(gt_entities))\n",
    "    # FP = len(set(ans_entities) - set(gt_entities))\n",
    "    # FN = len(set(gt_entities) - set(ans_entities))\n",
    "    # confusion_matrix[\"test\"][0] += TP\n",
    "    # confusion_matrix[\"test\"][1] += FP\n",
    "    # confusion_matrix[\"test\"][2] += FN\n",
    "    # f1_score = []\n",
    "    # # for category in categories:\n",
    "    # TP, FP, FN = confusion_matrix[\"test\"]\n",
    "    # precision = TP/(TP+FP) if TP+FP else float('nan')\n",
    "    # recall = TP/(TP+FN) if TP+FN else float('nan')\n",
    "    # f1_score.append(2*(precision * recall) / (precision + recall))\n",
    "    # return sum(f1_score) / len(f1_score)\n",
    "\n",
    "macro_avg_f1_score(entities_extracted, entities)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abbdb50e-bfda-4fdf-9bdf-47ef3a023efb",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
